libs/core/synchronization/include/hpx/synchronization/async_rw_mutex.hpp

```diff
@@ -0,0 +1,469 @@
+            using error_types = Variant<std::exception_ptr>;
+
+            static constexpr bool sends_done = false;
+
+            template <typename R>
+            struct operation_state
+            {
+                std::decay_t<R> r;
+                std::shared_ptr<readwrite_type> value;
+                std::shared_ptr<detail::async_rw_mutex_shared_state> prev_state;
+                std::shared_ptr<detail::async_rw_mutex_shared_state> state;
+
+                operation_state(operation_state&&) = delete;
+                operation_state(operation_state const&) = delete;
+
+                void start() noexcept
+                {
+                    HPX_ASSERT_MSG(prev_state,
+                        "async_rw_lock::sender::operation_state prev_state is "
+                        "empty, was the sender already started?");
+                    HPX_ASSERT_MSG(state,
+                        "async_rw_lock::sender::operation_state state is "
+                        "empty, was the sender already started?");
+
+                    prev_state->add_continuation(
+                        [r = std::move(r), value = std::move(value),
+                            state = std::move(state)]() mutable {
+                            try
+                            {
+                                hpx::execution::experimental::set_value(
+                                    std::move(r),
+                                    detail::async_rw_mutex_access_wrapper<
+                                        readwrite_type, read_type, AccessType>{
+                                        std::move(value), std::move(state)});
+                            }
+                            catch (...)
+                            {
+                                hpx::execution::experimental::set_error(
+                                    std::move(r), std::current_exception());
+                            }
+                        });
+
+                    // We release prev_state here to allow continuations to run.
+                    // The operation state may otherwise keep it alive longer
+                    // than needed.
+                    prev_state.reset();
+                }
+            };
+
+            template <typename R>
+            auto connect(R&& r) &&
+            {
+                return operation_state<R>{std::forward<R>(r), std::move(value),
+                    std::move(prev_state), std::move(state)};
+            }
+        };
+
+        std::shared_ptr<readwrite_type> value;
+        Allocator alloc;
+
+        detail::async_rw_mutex_access_type prev_access =
+            detail::async_rw_mutex_access_type::readwrite;
+
+        std::shared_ptr<detail::async_rw_mutex_shared_state> prev_state;
+        std::shared_ptr<detail::async_rw_mutex_shared_state> state =
+            std::allocate_shared<detail::async_rw_mutex_shared_state>(alloc);
+    };
+}}    // namespace hpx::experimental
```

Comment:
General question: all of the types defined here are move-only at best. Would using `unique_ptr` instead of `shared_ptr` be an option? And if yes, do we have to use a pointer type in the first place?

Reply:
Read-only access requires the shared state to be a `shared_ptr`. Even in the read-write case, while only one sender sends the shared state in the wrapper, the following senders need to hold on to the same shared state to attach their continuations as well.

Reply:
That said, I realize now that the value could be in a `unique_ptr` instead of a `shared_ptr`. The reference wrappers can hold references rather than `shared_ptr`s and the last continuation can release the `unique_ptr`. Now I'll have to reexamine if the shared states can do the same (I don't think so).

Reply:
If using a `unique_ptr<foo>` is possible, wouldn't be a plain `foo` possible as well? This should be viable at least for moveable or trivially copyable value types, shouldn't it?

Reply:
Maybe you see something I don't. I want heap allocation to have the value live past the end of the `async_rw_mutex` destructor (I'd prefer not to block there).

Reply:
Don't pay attention to what I said, I was merely trying to understand the code base.

Reply:
I will do exactly the opposite, thanks for looking through this!

Reply:
I've updated it to use a `unique_ptr` for the guarded value now.